odl/discr/lp_discr.py

```diff
@@ -164,56 +228,82 @@ def element(self, inp=None):
-        csize[idcs] = self.domain.extent[idcs]
-        return csize
-
-    @property
-    def cell_volume(self):
-        """Cell volume of an underlying regular grid."""
-        return float(np.prod(self.cell_size))
-
-    def points(self):
-        """All points in the sampling grid."""
-        return self.grid.points(order=self.order)
-
-    @property
-    def order(self):
-        """Axis ordering for array flattening."""
-        return self._order
-
-    @property
     def interp(self):
         """Interpolation type of this discretization."""
         return self._interp
 
+    # Overrides for space functions depending on partition
+    def _inner(self, x, y):
+        """Return ``self.inner(x, y)``."""
+        if any(bdry[0] or bdry[1]
+               for bdry in self.partition.nodes_on_boundary()):
+
+            # Need to halve the boundary nodes to get correct contributions.
+            # This requires copies, unfortunately.
+            # TODO: implement without copy
+            x_cpy = x.copy()
+            on_bdry = self.partition.nodes_on_boundary()
+            apply_on_boundary(
+                x_cpy.asarray(), func=lambda x: x / 2,
+                only_once=False, which_boundaries=on_bdry)
+
+            return super()._inner(x_cpy, y)
+        else:
+            # Standard case
+            return super()._inner(x, y)
+
+    def _norm(self, x):
+        """Return ``self.norm(x)``."""
+        if any(bdry[0] or bdry[1]
+               for bdry in self.partition.nodes_on_boundary()):
+
+            # TODO: implement without copy
+            x_cpy = x.copy()
+            on_bdry = self.partition.nodes_on_boundary()
+            if self.exponent != float('inf'):
+                # For p != inf we mimic the integral of f(x)**p, i.e. by
+                # the following scaling we effectively multiply the
+                # boundary contribution by 1/2.
+                bdry_fac = 2 ** (1.0 / self.exponent)
+                apply_on_boundary(
+                    x_cpy.asarray(), func=lambda x: x / bdry_fac,
+                    only_once=False, which_boundaries=on_bdry)
+
+            return super()._norm(x_cpy)
+        else:
+            # Standard case
+            return super()._norm(x)
+
+    def _dist(self, x, y):
+        """Return ``self.dist(x, y)``."""
+        if any(bdry[0] or bdry[1]
+               for bdry in self.partition.nodes_on_boundary()):
+
+            # TODO: implement without copy
+            x_cpy, y_cpy = x.copy(), y.copy()
+            on_bdry = self.partition.nodes_on_boundary()
+            if self.exponent != float('inf'):
+                # For p != inf we mimic the integral of (f(x)-g(x)**p, i.e. by
+                # the following scaling we effectively multiply the
+                # boundary contribution by 1/2.
+                bdry_fac = 2 ** (1.0 / self.exponent)
+                apply_on_boundary(
+                    x_cpy.asarray(), func=lambda x: x / bdry_fac,
```

Comment:
There would be a copy in that case right? Needs to write back then. Context manager would be cool.